### ISSUE TITLE
Apparent thread-safety issue in tensorflow/core/common_runtime/executor.cc

### ISSUE THREAD

**Issue by dtrebbien** [tensorflow/tensorflow#17116]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS 10.13.3
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: ('v1.6.0-rc1-277-g993006fa76', '1.6.0-rc1')
- **Python version**: Python 2.7.14
- **Bazel version (if compiling from source)**: 0.10.1-homebrew
- **GCC/Compiler version (if compiling from source)**: 4.2.1
- **CUDA/cuDNN version**: 9.1 / 7.0.5
- **GPU model and memory**: NVIDIA GeForce GT 750M with 2 GB device memory (CUDA compute capability 3.0)
- **Exact command to reproduce**: N/A

### Describe the problem
Clang warns about a thread-safety issue in `tensorflow/core/common_runtime/executor.cc` at line 2338 which warning appears to be valid.

Here is the code around that line:

```c++
  if (parent_frame != nullptr) {
    mutex_lock paranet_frame_lock(parent_frame->mu);
    // Propagate all the dead exits to the parent frame.
    for (const Node* node : frame->dead_exits) {                         // Line 2338
      auto parent_iter_state = parent_frame->GetIteration(parent_iter);
```

A lock is held on `parent_frame->mu`, but not `frame->mu`.

If there is no thread safety issue, I think that a comment should be added to explain why, as it's not clear.

### Source code / logs
<pre>
tensorflow/core/common_runtime/executor.cc:2338:27: warning: reading variable 'dead_exits' requires holding mutex 'frame->mu' [-Wthread-safety-precise]
    for (const Node* node : frame->dead_exits) {
                          ^
tensorflow/core/common_runtime/executor.cc:2338:27: note: found near match 'parent_frame->mu'
tensorflow/core/common_runtime/executor.cc:2338:27: warning: reading variable 'dead_exits' requires holding mutex 'frame->mu' [-Wthread-safety-precise]
    for (const Node* node : frame->dead_exits) {
                          ^
tensorflow/core/common_runtime/executor.cc:2338:27: note: found near match 'parent_frame->mu'
</pre>

**Comment by asimshankar** [comment]
@skye : Mind taking a look? (Since you might have been looking into `FrameState` lately?)

**Comment by skye** [comment]
@mrry do you know if this is intentional? I'll take a look when I get the chance if not.

**Comment by mrry** [comment]
It looks like that code is only called here on L2027:

https://github.com/tensorflow/tensorflow/blob/a4f1478134cdbf73f0ad7eda3e73407135a54566/tensorflow/core/common_runtime/executor.cc#L2024-L2027

I suspect that `is_frame_done` implies that there is only one thread operating on `frame`, but it doesn't look like any annotations or lock acquisitions protect that invariant.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @skye: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dtrebbien** [comment]
I think that this was fixed by 3d9c27742693f9859e2fb75de57fe108520de712
